File: 2021/2021W42/PreppindataW42.py
import pandas as pd

fundraise = pd.read_csv("C:\\Data\\PreppinData\\Prep Generate Rows datasets - Charity Fundraiser.csv", 
                          parse_dates=['Date'],
                          date_parser=lambda x: pd.to_datetime(x, format='%d/%m/%Y'))
fundraise.index = pd.DatetimeIndex(fundraise['Date'])
fundraise = fundraise.reindex(pd.date_range(fundraise.index.min(), fundraise.index.max()),
                                  method='ffill').rename_axis('FillDate').reset_index()
fundraise = fundraise.sort_values('FillDate').rename_axis('Days into fund raising').reset_index()
fundraise['Value raised per day'] = fundraise.apply(lambda x: x['Total Raised to date']/x['Days into fund raising'] 
                                                        if x['Days into fund raising']>0 else None, axis=1)
fundraise['Date'] = fundraise['FillDate'].dt.strftime('%A')
fundraise['Average raised per weekday'] = fundraise['Value raised per day'].groupby(fundraise['Date']).transform('mean')
fundraise = fundraise[['Date', 'Total Raised to date', 'Days into fund raising',
                       'Value raised per day', 'Average raised per weekday']]

File: 2021/2021W42/preppindataW42.R
library(readr)
library(dplyr)
library(tidyr)

Sys.setlocale("LC_ALL","English")

final <- read_csv("C:/Data/PreppinData/Prep Generate Rows datasets - Charity Fundraiser.csv",
                  col_types =cols(`Date` = col_date(format = "%d/%m/%Y"))) %>%
  complete(., `Date` = full_seq(`Date`, period = 1)) %>%
  arrange(`Date`) %>%
  fill(`Total Raised to date`) %>%
  mutate('Days into fund raising' = row_number()-1,
         'Value raised per day' = if_else(`Days into fund raising`>0, `Total Raised to date`/`Days into fund raising`, NA_real_), 
         'Date' = weekdays(`Date`)) %>%
  group_by(`Date`) %>%
  mutate('Average raised per weekday' = mean(`Value raised per day`, na.rm=TRUE))

View(final)
